Add Custom Fields to Product Model in Networking

pmusolino opened the pull request:
Closes: #13751 part of #13493

## Summary description of changes

This PR introduces custom fields to the product model, improving the flexibility of metadata handling. The changes include updates to the product struct, encoding logic, and associated mock objects to accommodate these custom fields, which will be used later. Right now, the custom fields are not used in the code, and also the Storage/Yosemite layer implementation is missing and will be handled later on.

In future PRs, AddOns and subscription that currently use metadata, should be also refactored.

## Summary of changes
- Adds a `customFields` property to the `Product` model, allowing for additional metadata to be associated with products.
- Updates the encoding process to include the `customFields` when serializing the `Product` object.
- Modifies various files to ensure that `customFields` are initialized and handled correctly across the system.
- Introduces relevant TODO comments for future implementation of custom field handling within the storage layer.

## Testing
- I verified that products works like before.
- I did a smoke testing to understand if AddOns and Products with subscriptions works like before.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
Merge commit '531676ced86f77c774d51b2fe5aa15f0bd316ffa' into issue/13493-new-metadata-entity-in-product

pmusolino committed:
Merge commit '62aea4b3c9b99b68ba76e96a82305a93d49ae1f1' into issue/13493-new-metadata-entity-in-product

pmusolino committed:
Merge commit 'e3950f96be572823e2bd24b3ac86a7fbb61752c7' into issue/13493-new-metadata-entity-in-product

pmusolino committed:
Merge commit '95424625fe659dba8b21ba50b51b23bbc4c8b3a8' into issue/13493-new-metadata-entity-in-product

pmusolino committed:
Merge commit '7481aca0dfc5fa99deb51abfd70d4dfd1b4b6b61' into issue/13493-new-metadata-entity-in-product

pmusolino committed:
Add custom fields to product models and handling

- Update `Networking.generated.swift` to include `customFields` in `Networking.Product`.
- Modify `Models+Copiable.generated.swift` to add `customFields` parameter in product methods.
- Adjust `AIProduct.swift` to initialize `customFields` as an empty array.
- Enhance `Product.swift` to declare and handle `customFields`.
- Update `Product+SwiftUIPreviewHelpers.swift` to include `customFields` in product initialization.
- Modify `ProductFactory.swift` to initialize products with empty `customFields`.
- Add `customFields` to mock product in `MockObjectGraph.swift`.
- Adjust `Product+ReadOnlyConvertible.swift` to handle `customFields` in product storage.

pmusolino committed:
update: added a TODO for encoding Custom Fields

pmusolino committed:
update: comment

pmusolino committed:
Merge commit 'c2f38422ddb12b891afbab9962869affa7722af2' into issue/13493-new-metadata-entity-in-product

selanthiraiyan commented on Networking/Networking/Model/Product/Product.swift:
```diff
@@ -521,6 +525,9 @@ public struct Product: Codable, GeneratedCopiable, Equatable, GeneratedFakeable
 
         let menuOrder = try container.decode(Int.self, forKey: .menuOrder)
 
+        // Filter out metadata if the key is prefixed with an underscore (internal meta keys)
+        let customFields = (try? container.decode([MetaData].self, forKey: .metadata).filter({ !$0.key.hasPrefix("_")})) ?? []
```
I think we should add unit tests to ensure that `customFields` are mapped as expected. Probably to `ProductMapperTests`. What do you think?

pmusolino replied:
Good point. I implemented this test: ff9b457560f3ea7ff728b7b8ce0ac01115da2cd6

selanthiraiyan reviewed (commented):
I smoke testes products tab in general, and it works as before. ✅ 

I think we should add unit tests to ensure that `customFields` are parsed as expected from a mock JSON. Please correct me if I am missing any context here.

pmusolino committed:
feat: add tests for custom fields in ProductMapper

- Implement test_custom_fields_are_properly_parsed in ProductMapperTests.swift to verify custom fields parsing.
- Update product.json to include a new custom field with id 6000, key "just_a_custom_field", and value "10".

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/222">20.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan reviewed (approved):
Thanks for adding the unit test, Paolo! The code looks good to me. ✅ 

The PR is ready to merge once we fix that failing unit test. 🚢
